execute/tokendata/usdc/usdc.go

```diff
@@ -4,43 +4,168 @@ import (
+) *USDCCCTPTokenDataObserver {
+	supportedTokens := make(map[string]struct{})
+	for chainSelector, tokenConfig := range configs.Tokens {
+		key := sourceTokenIdentifier(chainSelector, tokenConfig.SourcePoolAddress)
+		supportedTokens[key] = struct{}{}
+	}
+
+	return &USDCCCTPTokenDataObserver{
+		configs:           configs,
+		supportedTokens:   supportedTokens,
+		usdcMessageReader: usdcMessageReader,
+		attestationClient: attestationClient,
+	}
 }
 
 func (u *USDCCCTPTokenDataObserver) Observe(
 	ctx context.Context,
 	messages exectypes.MessageObservations,
 ) (exectypes.TokenDataObservations, error) {
-	// Ask LogPoller for USDC messages and matching events
-	usdcMessages, err := u.pickOnlyUSDCMessages(messages)
+	// 1. Pick only messages that contain USDC tokens
+	usdcMessages := u.pickOnlyUSDCMessages(messages)
+
+	// 2. Fetch USDC messages hashes based on the `MessageSent (bytes message)` event
+	usdcMessageHashes, err := u.fetchUSDCMessageHashes(ctx, usdcMessages)
 	if err != nil {
 		return nil, err
 	}
 
-	// Fetch attestations for USDC messages
-	attestations, err := u.fetchAttestations(ctx, usdcMessages)
+	// 3. Fetch attestations for USDC messages
+	attestations, err := u.fetchAttestations(ctx, usdcMessageHashes)
 	if err != nil {
 		return nil, err
 	}
 
-	// Add attestations to the token observations
-	return u.extractTokenData(attestations)
+	// 4. Add attestations to the token observations
+	return u.extractTokenData(messages, attestations)
 }
 
-func (u *USDCCCTPTokenDataObserver) pickOnlyUSDCMessages(_ exectypes.MessageObservations) (interface{}, error) {
-	fmt.Println(u.configs)
-	panic("implement me")
+func (u *USDCCCTPTokenDataObserver) IsTokenSupported(
+	sourceChain cciptypes.ChainSelector,
+	msgToken cciptypes.RampTokenAmount,
+) bool {
+	_, ok := u.supportedTokens[sourceTokenIdentifier(sourceChain, msgToken.SourcePoolAddress.String())]
+	return ok
+}
+
+func (u *USDCCCTPTokenDataObserver) pickOnlyUSDCMessages(
+	messageObservations exectypes.MessageObservations,
+) map[cciptypes.ChainSelector]map[cciptypes.SeqNum][]int {
+	usdcMessages := make(map[cciptypes.ChainSelector]map[cciptypes.SeqNum][]int)
+	for chainSelector, messages := range messageObservations {
+		usdcMessages[chainSelector] = make(map[cciptypes.SeqNum][]int)
+		for seqNum, message := range messages {
+			var usdcTokens []int
+			for i, tokenAmount := range message.TokenAmounts {
+				tokenIdentifier := sourceTokenIdentifier(chainSelector, tokenAmount.SourcePoolAddress.String())
+				if _, ok := u.supportedTokens[tokenIdentifier]; ok {
+					usdcTokens = append(usdcTokens, i)
+				}
+			}
+			if len(usdcTokens) > 0 {
+				usdcMessages[chainSelector][seqNum] = usdcTokens
+			}
+		}
+	}
+	return usdcMessages
+}
+
+func (u *USDCCCTPTokenDataObserver) fetchUSDCMessageHashes(
+	ctx context.Context,
+	usdcMessages map[cciptypes.ChainSelector]map[cciptypes.SeqNum][]int,
+) (map[cciptypes.ChainSelector]map[cciptypes.SeqNum]map[int][]byte, error) {
```

Comment:
I wonder if some "key flattening" would help here now that you have 3 nested maps? Maybe instead the type can be:

```suggestion
type seqNumTokIdx struct {
  seqNum cciptypes.SeqNum
  // tokIdx is the index into the token amounts array part of the message
  tokIdx int
}

) (map[cciptypes.ChainSelector]map[seqNumTokIdx][]byte, error) {
```

Reply:
Having just the `int` type in the map keys feels pretty opaque to me. Similar thoughts on the other 3x nested map types

Reply:
These nested maps are terrible indeed. I want to reiterate that once I start implementing that interface